File: starrysky/src/main/java/com/lzx/starrysky/provider/MediaResource.java
package com.lzx.starrysky.provider;

import android.text.TextUtils;

import java.util.HashMap;

/**
 * 播放媒体包装
 */
public class MediaResource {
    private String mediaId;
    private long queueId;
    private String mediaUrl;
    private HashMap<String, MediaResource> mCacheMediaResource = new HashMap<>();

    private MediaResource(String mediaId, String mediaUrl, long queueId) {
        this.mediaId = mediaId;
        this.mediaUrl = mediaUrl;
        this.queueId = queueId;
    }

    public MediaResource() {
    }

    public MediaResource obtain(String mediaId, String mediaUrl, long queueId) {
        if (!TextUtils.isEmpty(mediaId)) {
            MediaResource resource = mCacheMediaResource.get(mediaId);
            if (resource == null) {
                resource = new MediaResource(mediaId, mediaUrl, queueId);
                mCacheMediaResource.put(mediaId, resource);
            }
            if (!resource.getMediaUrl().equals(mediaUrl)) {
                resource.mediaUrl = mediaUrl;
            }
            return resource;
        } else {
            throw new IllegalStateException("songId is null");
        }
    }

    public String getMediaId() {
        return mediaId;
    }

    public long getQueueId() {
        return queueId;
    }

    public String getMediaUrl() {
        return mediaUrl;
    }
}

File: starrysky/src/main/java/com/lzx/starrysky/provider/MediaResource.kt
package com.lzx.starrysky.provider

class MediaResource constructor() {

    private lateinit var mediaId: String
    private var queueId: Long = 0L
    private lateinit var mediaUrl: String
    private var mMapHeadData: Map<String, String>? = hashMapOf()
    private var mCacheMediaResource = hashMapOf<String, MediaResource>()

    constructor(
        mediaId: String,
        mediaUrl: String,
        queueId: Long,
        headData: Map<String, String>?
    ) : this() {
        this.mediaId = mediaId
        this.queueId = queueId
        this.mediaUrl = mediaUrl
        this.mMapHeadData = headData
    }

    fun obtain(
        mediaId: String?, mediaUrl: String, queueId: Long, headData: Map<String, String>?
    ): MediaResource {
        if (!mediaId.isNullOrEmpty()) {
            var resource = mCacheMediaResource[mediaId]
            if (resource == null) {
                resource = MediaResource(mediaId, mediaUrl, queueId, headData)
                mCacheMediaResource[mediaId] = resource
            }
            if (resource.getMediaUrl() != mediaUrl) {
                resource.mediaUrl = mediaUrl
            }
            return resource
        } else {
            throw IllegalStateException("songId is null")
        }
    }

    fun getMediaId(): String? {
        return mediaId
    }

    fun getQueueId(): Long {
        return queueId
    }

    fun getMediaUrl(): String? {
        return mediaUrl
    }

    fun getMapHeadData(): Map<String, String>? {
        return mMapHeadData
    }
}